### DISCUSSION TITLE
(SEO) Tổng hợp về kiến thức và kinh nghiệm khi làm việc với SEO một cách hiệu quả và an toàn

### DISCUSSION THREAD

**Discussion by lamminhthien** [abc-software-solutions-company/todo-list-frontend#626] [Code Library]
# Hiểu về SEO và Cách Các Thẻ Meta Ảnh Hưởng Đến Tối Ưu Hóa Trang Web

Trong môi trường kỹ thuật số ngày nay, tối ưu hóa công cụ tìm kiếm (SEO - Search Engine Optimization) là một yếu tố quan trọng giúp website của bạn xuất hiện nổi bật trên các trang kết quả tìm kiếm. Một trong những yếu tố cơ bản nhưng quan trọng nhất trong SEO chính là các thẻ meta. Thông qua bài viết này, chúng ta sẽ tìm hiểu chi tiết về cách các thẻ meta hoạt động và ảnh hưởng đến SEO.

## SEO là gì?

SEO (Search Engine Optimization) là quá trình tối ưu hóa trang web của bạn để nó có thể xuất hiện ở vị trí cao trên các trang kết quả tìm kiếm, như Google. Mục tiêu chính của SEO là tăng cường lưu lượng truy cập từ các công cụ tìm kiếm thông qua việc cải thiện thứ hạng trang web.

## Các thẻ Meta là gì?

Thẻ meta là các phần tử trong mã nguồn HTML của trang web, chứa thông tin mô tả về nội dung của trang web cho các công cụ tìm kiếm và trình duyệt. Mặc dù không hiển thị trực tiếp trên trang web, nhưng chúng có vai trò quan trọng trong việc ảnh hưởng đến cách trang web của bạn được hiểu và xếp hạng bởi các công cụ tìm kiếm.

## Phân tích chi tiết các thẻ Meta

Hãy xem qua một ví dụ cụ thể về các thẻ meta và cách chúng ảnh hưởng đến SEO của trang web.

### 1. **Viewport và Charset**

```html
<meta name="viewport" content="width=device-width, initial-scale=1" /> <meta charset="utf-8" />
```

- **Viewport**: Thẻ này xác định cách trang web của bạn hiển thị trên các thiết bị khác nhau, đặc biệt là trên di động. Điều này cực kỳ quan trọng vì Google ưu tiên các trang web thân thiện với thiết bị di động trong việc xếp hạng.
- **Charset**: Thiết lập mã hóa ký tự của trang web, thường là `utf-8`, giúp đảm bảo rằng các ký tự đặc biệt được hiển thị đúng cách.

### 2. **Title và Description**

```html
<title>Home | Your Website</title> <meta name="description" content="Home Description" />
```

- **Title**: Thẻ tiêu đề là một trong những yếu tố SEO quan trọng nhất. Đây là dòng tiêu đề xuất hiện trên trang kết quả tìm kiếm và ảnh hưởng trực tiếp đến quyết định của người dùng khi chọn nhấp vào liên kết. Tiêu đề nên ngắn gọn, chứa từ khóa chính và hấp dẫn người đọc.
- **Description**: Thẻ mô tả cung cấp một tóm tắt ngắn gọn về nội dung trang web. Dù không trực tiếp ảnh hưởng đến thứ hạng SEO, nhưng một mô tả hấp dẫn có thể tăng tỷ lệ nhấp (CTR), góp phần gián tiếp vào SEO.

### 3. **Thẻ Tác giả và Nhà xuất bản**

```html
<meta name="author" content="Your Company" /> <meta name="publisher" content="Your Company" />
```

- **Author & Publisher**: Thẻ này cho biết ai là người tạo nội dung và xuất bản trang web. Điều này có thể giúp xây dựng uy tín và niềm tin từ cả người dùng và công cụ tìm kiếm, đặc biệt nếu trang web của bạn liên quan đến các nội dung chuyên môn hoặc cần độ tin cậy cao.

### 4. **Thẻ từ khóa (Keywords)**

```html
<meta name="keywords" content="travel, tour, booking, trip, vietnam, nha trang" />
```

- **Keywords**: Trước đây, thẻ từ khóa được sử dụng phổ biến để chỉ định các từ khóa chính cho trang web. Tuy nhiên, ngày nay các công cụ tìm kiếm lớn như Google không còn sử dụng thẻ này trong việc xếp hạng trang web do lạm dụng từ khóa.

### 5. **Canonical**

```html
<link rel="canonical" href="https://www.example.com/" />
```

- **Canonical**: Thẻ này giúp ngăn chặn tình trạng trùng lặp nội dung bằng cách chỉ định URL chuẩn cho trang. Điều này đặc biệt hữu ích khi có nhiều phiên bản của một trang tồn tại và bạn muốn công cụ tìm kiếm chỉ tập trung vào một phiên bản duy nhất.

### 6. **Thẻ Open Graph (OG)**

```html
<meta property="og:title" content="Your Website | Your Title" />
<meta property="og:description" content="Your description goes here." />
<meta property="og:url" content="https://www.example.com/" />
<meta property="og:site_name" content="Your Website" />
<meta property="og:image" content="https://www.example.com/og-img.jpg" />
<meta property="og:image:width" content="1200" />
<meta property="og:image:height" content="680" />
<meta property="og:type" content="website" />
```

- **Open Graph**: Các thẻ Open Graph được sử dụng chủ yếu để tối ưu hóa cách mà nội dung của bạn hiển thị khi được chia sẻ trên các nền tảng mạng xã hội như Facebook, LinkedIn. Chúng giúp định hình cách hiển thị của bài viết với các yếu tố như tiêu đề, mô tả và hình ảnh.

### 7. **Thẻ Twitter Card**

```html
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:site" content="@site" />
<meta name="twitter:creator" content="@creator" />
<meta name="twitter:title" content="Your Website | Your Title" />
<meta name="twitter:description" content="Your description goes here." />
```

- **Twitter Card**: Tương tự như Open Graph, các thẻ này tối ưu hóa cách nội dung của bạn hiển thị trên Twitter, bao gồm tiêu đề, mô tả và hình ảnh.

## Ví dụ về cách tích hợp SEO vào Next.JS App Router với file layout.tsx

```js
import { ReactNode } from "react";
import { NextIntlClientProvider } from "next-intl";
import { notFound } from "next/navigation";
import { Metadata } from "next";

// Define supported locales
const locales = ["en", "es", "fr"];

// Define types for Layout props
type LayoutProps = {
  children: ReactNode,
  params: { locale: string },
};

// Site configuration
const siteConfigs = {
  url: "https://www.globaltrek.com",
  appName: "GlobalTrek",
  appSlogan: "Explore the World",
  description: "Your ultimate travel companion app",
  keyword: "travel, adventure, explore, global, app",
  companyName: "World Explorers Ltd.",
  author: { name: "Jane Traveler" },
  appNameSlogan: "GlobalTrek - Explore the World",
};

export default async function Layout({ children, params: { locale } }: LayoutProps) {
  // Define the environment "stage" | "preprod" | "production"
  const environment = process.env.NEXT_PUBLIC_ENVIRONMENT;

  // Check if the locale is valid
  if (!locales.includes(locale)) notFound();

  // Load messages for the locale
  let messages;
  try {
    messages = (await import(`@/messages/${locale}.json`)).default;
  } catch (error) {
    notFound();
  }

  return (
    <html lang={locale}>
      <body>
        <NextIntlClientProvider locale={locale} messages={messages}>
          {children}
        </NextIntlClientProvider>
      </body>
    </html>
  );
}

// Generate static params for locales
export function generateStaticParams() {
  return locales.map((locale) => ({ locale }));
}

// Generate metadata with specified properties
export function generateMetadata({ params: { locale } }: LayoutProps): Metadata {
  const url = `${siteConfigs.url}/${locale}`;

  return {
    metadataBase: new URL(siteConfigs.url || "https://www.globaltrek.com"),
    title: {
      default: siteConfigs.appName + " | " + siteConfigs.appSlogan,
      template: "%s | " + siteConfigs.appName,
    },
    description: siteConfigs.description,
    applicationName: siteConfigs.appName,
    keywords: siteConfigs.keyword,
    creator: siteConfigs.companyName,
    publisher: siteConfigs.companyName,
    authors: [{ name: siteConfigs.author.name, url: "https://worldexplorers.com/" }],
    twitter: {
      title: siteConfigs.appNameSlogan,
      card: "summary_large_image",
      site: "@GlobalTrek",
      creator: "@JaneTraveler",
      images: `${siteConfigs.url}/og-img.jpg`,
    },
    openGraph: {
      url,
      siteName: siteConfigs.appName,
      title: siteConfigs.appNameSlogan,
      description: siteConfigs.description,
      type: "website",
      images: [{ alt: siteConfigs.appName, url: `${siteConfigs.url}/og-img.jpg`, width: 1200, height: 680 }],
    },
    colorScheme: "light",
    themeColor: [
      { media: "(prefers-color-scheme: dark)", color: "#1a365d" },
      { media: "(prefers-color-scheme: light)", color: "#e6f2ff" },
    ],
    viewport: "width=device-width, initial-scale=1",
    manifest: "/manifest.json",
    icons: [
      { rel: "shortcut icon", type: "image/x-icon", url: "/favicon.ico" },
      { rel: "icon", type: "image/x-icon", url: "/favicon.ico" },
      { rel: "icon", sizes: "16x16", type: "image/png", url: "/favicon-16x16.png" },
      { rel: "icon", sizes: "32x32", type: "image/png", url: "/favicon-32x32.png" },
      { rel: "apple-touch-icon", sizes: "180x180", url: "/apple-touch-icon.png" },
      { rel: "mask-icon", color: "#5bbad5", url: "/safari-pinned-tab.svg" },
    ],
    robots:
      environment === "stage" || environment === "preprod"
        ? { index: false, follow: false }
        : { index: true, follow: true },
    other: {
      HandheldFriendly: "true",
      MobileOptimized: "360",
      google: "notranslate",
    },
    alternates: {
      canonical: url,
    },
  };
}
```

## Khi cần cá nhân hoá nội dung SEO cho từng trang riêng lẻ, ta có thể ghi đè một vài thuộc tính SEO mặc định trong file page.tsx đối với Next.JS App Router:

```js
// src/app/[locale]/(auth)/login/page.tsx
import { Metadata } from "next";

export interface IBasePageProps {
  params: { locale: string };
  searchParams: { [key: string]: string | string[] | undefined };
}

export default async function LoginPage() {
  return (
    <div>
      <h1>Login Page</h1>
    </div>
  );
}

export async function generateMetadata({ params: { locale } }: IBasePageProps): Promise<Metadata> {
  return {
    title: "Login",
    description: "Login Description",
    alternates: {
      canonical: `/${locale}/login`,
    },
  };
}
```

## Tầm quan trọng của việc tối ưu hóa thẻ Meta

Tối ưu hóa các thẻ meta không chỉ giúp cải thiện thứ hạng SEO mà còn giúp trang web của bạn dễ dàng tiếp cận với người dùng hơn. Thông qua việc cung cấp thông tin chính xác và hấp dẫn, bạn có thể tăng cường trải nghiệm người dùng, nâng cao tỷ lệ nhấp và cuối cùng là đạt được các mục tiêu kinh doanh của mình.

Kết luận, SEO không chỉ đơn thuần là việc áp dụng từ khóa mà còn bao gồm việc tối ưu hóa toàn diện các yếu tố khác nhau trên trang web, trong đó các thẻ meta đóng vai trò quan trọng. Hãy đảm bảo rằng trang web của bạn có các thẻ meta được tối ưu hóa để đạt được hiệu quả SEO tốt nhất.

# SEO An Toàn: Chặn Kết Quả Tìm Kiếm ở Môi Trường Không Phải Production

Khi phát triển web, các môi trường như staging, preprod cần được bảo vệ để tránh nội dung chưa hoàn thiện bị lập chỉ mục trên Google. Điều này giúp tránh trùng lặp nội dung và bảo vệ SEO cho môi trường production.

## Sử Dụng Thẻ Meta Để Chặn Google Bot

Sử dụng thẻ meta sau để ngăn Google lập chỉ mục và theo dõi các liên kết trên môi trường không production:

- **noindex**: Ngăn trang xuất hiện trong kết quả tìm kiếm.
- **nofollow**: Ngăn phân phối "link juice" từ các liên kết trên trang này.

```html
<meta name="robots" content="noindex, nofollow" />
```

Tích hợp vào Next.JS, apply vào file layout.ts global

```js
export async function generateMetadata() {
  const environment = process.env.NEXT_PUBLIC_ENVIRONMENT;

  return {
    title: 'Example Page',
    robots: environment === 'stage' || environment === 'preprod' ? {index: false, follow: false}: {index: true, follow: true},
  };
}

export default function ExamplePage() {
  return (
    <div>
      <h1>Example Page</h1>
      <p>Nội dung trang này sẽ thay đổi metadata dựa trên môi trường.</p>
    </div>
  );
}
**

```

## Tích Hợp Vào Quy Trình Phát Triển Phần Mềm

- **Tự động thêm**: Sử dụng script hoặc biến môi trường để thêm thẻ trên vào các trang staging, preprod
- **Kiểm tra trước khi lên production**: Đảm bảo gỡ bỏ thẻ trước khi đưa lên production.

## Kết Luận

Bảo vệ SEO bằng cách chặn lập chỉ mục các trang không phải production, chỉ để nội dung hoàn thiện xuất hiện trên Google.

# Những Điều Cần Lưu Ý Khi Đưa Website Lên Live Và Bắt Đầu SEO

Khi chuẩn bị đưa website lên live và thực hiện SEO, cần chú ý các điểm sau để đảm bảo hiệu quả SEO và tránh sai sót:

## 1. Site Name (Tên Trang Web)

Trước khi đưa website lên production, đảm bảo chỉnh sửa thẻ meta `<meta property="og:site_name" content="Your Site Name">`. Thẻ này định nghĩa tên trang khi nội dung được chia sẻ trên mạng xã hội. Đôi khi, nếu sử dụng template, bạn có thể quên thay đổi tên trang web mặc định, dẫn đến thông tin không chính xác khi chia sẻ.

## 2. sitemap.xml

**sitemap.xml** giúp công cụ tìm kiếm hiểu cấu trúc website và dễ dàng crawl toàn bộ nội dung. Ví dụ:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
    <url>
        <loc>https://www.example.com/</loc>
        <lastmod>2024-09-04</lastmod>
        <priority>1.00</priority>
    </url>
    <url>
        <loc>https://www.example.com/blog/</loc>
        <lastmod>2024-09-03</lastmod>
        <priority>0.80</priority>
    </url>
</urlset>
```

- **loc**: Đường dẫn trang.
- **lastmod**: Ngày cập nhật cuối cùng.
- **priority**: Mức độ ưu tiên của trang đối với crawl.

### Ví dụ về sitemap.xml (sitemap cha)

```xml
<?xml version="1.0" encoding="UTF-8"?>
<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
    <sitemap>
        <loc>https://www.example.com/sitemap-pages.xml</loc>
        <lastmod>2024-09-04</lastmod>
    </sitemap>
    <sitemap>
        <loc>https://www.example.com/sitemap-blog.xml</loc>
        <lastmod>2024-09-03</lastmod>
    </sitemap>
</sitemapindex>
```

- **loc**: Đường dẫn đến file sitemap con.
- **lastmod**: Ngày cập nhật cuối cùng của sitemap con.

### Ví dụ về sitemap-1.xml (sitemap con)

```xml
<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
    <url>
        <loc>https://www.example.com/page1</loc>
        <lastmod>2024-09-01</lastmod>
        <priority>0.80</priority>
    </url>
    <url>
        <loc>https://www.example.com/page2</loc>
        <lastmod>2024-09-02</lastmod>
        <priority>0.80</priority>
    </url>
</urlset>
```

---

## Ví dụ về cách tích hợp sitemap.xml trong Next.JS App Router

Dưới đây là một ví dụ về cách tạo sitemaps động trong Next.js mà không liên quan đến quiz. Trong ví dụ này, chúng ta sẽ giả định việc lấy dữ liệu từ một API của trang web `example.com` để tạo các sitemaps.

### 1. Cấu hình Biến Môi Trường và URL Website

Đầu tiên, định nghĩa `WEBSITE_URL` trong file `constants.ts`:

```javascript
// src/common/constants/seo.constant.ts
export const WEBSITE_URL = process.env.NEXT_PUBLIC_WEBSITE_URL || "https://www.example.com";
```

Sau đó, trong file `.env`, bạn cần cấu hình URL trang web:

```bash
# .env
NEXT_PUBLIC_WEBSITE_URL=https://www.example.com
```

### 2. Tạo Sitemap Index

Để tạo sitemap index, bạn cần tạo file trong thư mục `app` (sử dụng App Router). Đây là nơi bạn sẽ nhóm nhiều sitemaps nhỏ thành một sitemap chính.

#### Cấu trúc File Sitemap Index

```javascript
// src/app/(sitemap)/sitemap.xml/route.ts
import { getServerSideSitemapIndex } from "next-sitemap";
import { WEBSITE_URL } from "@/common/constants";

export async function GET() {
  if (process.env.NODE_ENV === "production") {
    const listSitemap = Array.from({ length: 7 }, (_, index) => `${WEBSITE_URL}/sitemap-${index}.xml`);
    return getServerSideSitemapIndex(listSitemap);
  }
}
```

### 3. Tạo Sitemap Động cho Các Trang Khác

Bây giờ, chúng ta sẽ tạo một sitemap động bằng cách lấy dữ liệu từ một API giả lập. Ví dụ này sẽ lấy danh sách các bài viết từ API và tạo sitemap cho các bài viết đó.

#### Cấu trúc File Sitemap Bài Viết

```javascript
// src/app/(sitemap)/sitemap-1.xml/route.ts
import { getServerSideSitemap, ISitemapField } from "next-sitemap";
import PostsApi from "@/common/api/posts.api";
import { WEBSITE_URL } from "@/common/constants";

export async function GET() {
  try {
    const listSitemapPosts: ISitemapField[] = [];

    // Giả lập việc gọi API để lấy dữ liệu bài viết
    const res = await PostsApi.getPostsForSitemap();

    const posts = res.data.data.filter((post) => post?.slug);

    // Tạo các liên kết cho bài viết
    posts.map((post) => listSitemapPosts.push({ loc: `${WEBSITE_URL}/posts/${post.slug}` }));

    return getServerSideSitemap(listSitemapPosts);
  } catch {
    return getServerSideSitemap([]);
  }
}
```

### 4. Giả lập API

Dưới đây là cách giả lập một API trả về danh sách các bài viết từ `example.com`:

```javascript
// src/common/api/posts.api.ts
const PostsApi = {
  getPostsForSitemap: async () => {
    return {
      data: {
        data: [
          { slug: "post-1" },
          { slug: "post-2" },
          { slug: "post-3" },
          // Thêm các bài viết khác ở đây
        ],
      },
    };
  },
};

export default PostsApi;
```

### Kiểm tra Kết quả

Sau khi triển khai, bạn có thể kiểm tra sitemap bằng cách truy cập `https://www.example.com/sitemap-index.xml` hoặc `https://www.example.com/sitemap-posts.xml` để xem các URL đã được tạo tự động dựa trên dữ liệu từ API.

### Kết luận

Bằng cách sử dụng `next-sitemap` và tích hợp với API, bạn có thể tạo sitemaps động cho các trang web có nội dung thay đổi thường xuyên. Điều này giúp đảm bảo rằng các công cụ tìm kiếm luôn có thông tin cập nhật nhất về cấu trúc trang web của bạn, giúp tối ưu hóa SEO một cách hiệu quả.

## 3. robots.txt

**robots.txt** giúp kiểm soát phần nào của website được phép hoặc không được phép crawl bởi công cụ tìm kiếm. Ví dụ:

```txt
User-agent: *
Disallow: /admin/
Disallow: /protected/
Allow: /blog/
Sitemap: https://www.example.com/sitemap.xml
```

- **User-agent**: Áp dụng quy tắc cho tất cả công cụ tìm kiếm.
- **Disallow**: Ngăn không cho công cụ tìm kiếm truy cập các thư mục `/admin/` và `/protected/`.
- **Allow**: Cho phép crawl thư mục `/blog/`.
- **Sitemap**: Liên kết đến file sitemap.xml của website.

## Ví dụ về cách tính hợp file robot.txt trong Next.JS App Router

```js
// src/app/robots.ts
import { MetadataRoute } from "next";

export default function robots(): MetadataRoute.Robots {
  const WEBSITE_URL = process.env.NEXT_PUBLIC_SITE_URL;

  const listSitemap = Array.from({ length: 7 }, (_, index) => `${WEBSITE_URL}/sitemap-${index}.xml`);

  return {
    rules: [
      {
        userAgent: "*",
        allow: "/",
        disallow: ["/private/", "/protected/", "/admin/"],
      },
    ],
    host: WEBSITE_URL,
    sitemap: [`${WEBSITE_URL}/sitemap.xml`, ...listSitemap],
  };
}
```

**Comment by tintranfs** [comment]
@lamminhthien  Khi nói về một vấn đề em chia theo section và nên đưa ra giải pháp ngay bên dưới luôn. Tránh việc vấn đề ở trên rồi phải scroll xuống dưới thiệt dài mới xem được giải pháp.

**Reply by ace53thntu** [reply]
Đúng là hơi dài. 😄 scroll hơi mệt. chia ra thành child page hoặc markdown cũng có hỗ trợ collapse

**Reply by lamminhthien** [reply]
Về Collpase em rất thích tính năng này ở Notion.so
Em sẽ trải nghiệm bên Markdown thử.
